tools/java/org/unicode/cldr/util/XMLSource.java

```diff
@@ -1747,4 +1764,711 @@ public DtdType getDtdType() {
+            fis = InputStreamFactory.createInputStream(f);
+        } catch (IOException e) {
+            StringBuilder sb = new StringBuilder("Cannot read the file '");
+            sb.append(f);
+            throw new ICUUncheckedIOException(sb.toString(), e);
+        }
+
+        fis = new StripUTF8BOMInputStream(fis);
+        InputStreamReader reader = new InputStreamReader(fis, Charset.forName("UTF-8"));
+        XMLSource source = new SimpleXMLSource(localeId);
+        XMLNormalizingHandler XML_HANDLER = new XMLNormalizingHandler(source, minimalDraftStatus);
+        XMLFileReader.read(fullFileName, reader, -1, true, XML_HANDLER);
+        if (XML_HANDLER.isSupplemental < 0) {
+            throw new IllegalArgumentException("root of file must be either ldml or supplementalData");
+        }
+        source.setNonInheriting(XML_HANDLER.isSupplemental > 0);
+        if (XML_HANDLER.overrideCount > 0) {
+            throw new IllegalArgumentException("Internal problems: either data file has duplicate path, or" +
+                " CLDRFile.isDistinguishing() or CLDRFile.isOrdered() need updating: "
+                + XML_HANDLER.overrideCount
+                + "; The exact problems are printed on the console above.");
+        }
+        return source;
+    }
+
+    private static boolean LOG_PROGRESS = false;
+    private static final boolean DEBUG = false;
+
+    private static class XMLNormalizingHandler implements AllHandler {
+        private static UnicodeSet whitespace = new UnicodeSet("[:whitespace:]");
+        private DraftStatus minimalDraftStatus;
+        private static final boolean SHOW_START_END = false;
+        private int commentStack;
+        private boolean justPopped = false;
+        private String lastChars = "";
+        private String currentFullXPath = "/";
+        private String comment = null;
+        private Map<String, String> attributeOrder;
+        private DtdData dtdData;
+        private XMLSource source;
+        private String lastActiveLeafNode;
+        private String lastLeafNode;
+        private int isSupplemental = -1;
+        private int[] orderedCounter = new int[30]; // just make deep enough to handle any CLDR file.
+        private String[] orderedString = new String[30]; // just make deep enough to handle any CLDR file.
+        private int level = 0;
+        private int overrideCount = 0;
+
+        XMLNormalizingHandler(XMLSource source, DraftStatus minimalDraftStatus) {
+            this.source = source;
+            this.minimalDraftStatus = minimalDraftStatus;
+        }
+
+        private String show(Attributes attributes) {
+            if (attributes == null) return "null";
+            String result = "";
+            for (int i = 0; i < attributes.getLength(); ++i) {
+                String attribute = attributes.getQName(i);
+                String value = attributes.getValue(i);
+                result += "[@" + attribute + "=\"" + value + "\"]"; // TODO quote the value??
```

Comment:
Repeated String concatenation is slow, use StringBuilder.

Reply:
Agree and will change